In [1]:
!pip install statsmodels --quiet

In [2]:
import numpy as np
from statsmodels.stats.weightstats import ztest

В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

Завдання 1: Z-тест для однієї вибірки
Мета: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

Припущення: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній heights.

Для виконання завдання імпортуйте метод для проведення ztest, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [3]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [4]:
heights = np.array(heights)

In [5]:
len(heights)
np.mean(heights)

np.float64(170.6969696969697)

In [6]:
null_mean = 170
z_stat_1, p_value_1 = ztest(
    heights,
    value=170,
    alternative='larger'
)

In [7]:
z_stat_1, p_value_1

(np.float64(0.8482235371587619), np.float64(0.19815674571091751))

In [8]:
print(f"Z-statistic: {z_stat_1:.3f}")
print(f"P-value: {p_value_1:.3f}")

Z-statistic: 0.848
P-value: 0.198


In [9]:
alpha = 0.05

if p_value_1 < alpha:
    print("Відхиляємо H₀.")
else:
    print("Не відхиляємо H₀.")

Не відхиляємо H₀.


Ні, не можемо відхилити нульову гіпотезу H₀, оскільки p-value = 0.198 більше за рівень значущості α = 0.05. Припущення, що середній зріст студентів перевищує 170 см, на рівні значущості 0.05 статистично не підтвердилося.

Завдання 2: Z-тест для двох незалежних вибірок
Завдання 2.1.

Мета: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

Припущення: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [10]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [11]:
group_a_scores = np.array(group_a_scores)
group_b_scores = np.array(group_b_scores)

np.mean(group_a_scores), np.mean(group_b_scores)

(np.float64(77.08125), np.float64(74.73875000000001))

In [12]:
np.mean(group_a_scores), np.mean(group_b_scores)

(np.float64(77.08125), np.float64(74.73875000000001))

In [13]:
z_stat_21, p_value_21 = ztest(
    group_a_scores,
    group_b_scores,
    alternative='larger'
)

In [14]:
print(f"Z-statistic: {z_stat_21:.3f}")
print(f"P-value: {p_value_21:.3f}")

Z-statistic: 2.357
P-value: 0.009


In [15]:
if(p_value_21 <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


Відхиляємо нульову гіпотезу, оскільки p-value = 0.009 менше за рівень значущості α = 0.05. Є статистично значущі підстави вважати, що середній бал групи А вищий, ніж в групи B, і новий навчальний курс виявився ефективнішим.

Завдання 2.2.

Мета: Побачити, як впливає розмір вибірок на результат тесту.

Завдання: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше. Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний метод). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.


In [16]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [17]:
group_a_small = group_a_scores[:5]

In [18]:
group_a_small
len(group_a_small)

5

In [19]:
print(group_a_small)
print(len(group_a_small))

[78.55, 72.25, 79.88, 75.0, 76.54]
5


In [20]:
from scipy.stats import ttest_ind

In [21]:
t_stat_22, p_value_22 = ttest_ind(
    group_a_small,
    group_b_scores,
    alternative='greater'
)

In [22]:
print(f"T-statistic: {t_stat_22:.3f}")
print(f"P-value: {p_value_22:.3f}")

T-statistic: 0.817
P-value: 0.209


In [23]:
alpha = 0.05

if p_value_22 < alpha:
    print("Відхиляємо H₀.")
else:
    print("Не відхиляємо H₀.")

Не відхиляємо H₀.


Нульова гіпотеза не може бути відхилена, оскільки p-value = 0.209 > 0.05. За результатами 5 студентів недостатньо доказів, що новий курс ефективніший.

Завдання 3*: Z-тест для двох пов'язаних вибірок
Мета: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

Припущення: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [24]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [25]:
before_training = np.array(before_training)
after_training = np.array(after_training)

difference = after_training - before_training

In [26]:
difference
np.mean(difference)

np.float64(3.1499999999999995)

In [27]:
null_mean = 0

z_stat_3, p_value_3 = ztest(
    difference,
    value=null_mean,
    alternative='larger'
)

In [28]:
z_stat_3, p_value_3

(np.float64(9.13890009623916), np.float64(3.1545359919175505e-20))

In [29]:
print(f"Z-statistic: {z_stat_3:.3f}")
print(f"P-value: {p_value_3:.3e}")

Z-statistic: 9.139
P-value: 3.155e-20


In [30]:
alpha = 0.05

if(p_value_3 <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


Нульову гіпотезу відхиляємо, оскільки p-value менше за рівень значущості 0.05. Є статистично значущі підстави вважати, що показники співробітників після тренінгу покращилися.